src/Compilers/Core/Portable/CodeGen/PrivateImplementationDetails.cs

```diff
@@ -163,13 +166,33 @@ internal Cci.IFieldReference GetModuleVersionId(Cci.ITypeReference mvidType)
         {
             if (_mvidField == null)
             {
+                Debug.Assert(!IsFrozen);
                 Interlocked.CompareExchange(ref _mvidField, new ModuleVersionIdField(this, mvidType), null);
             }
 
             Debug.Assert(_mvidField.Type.Equals(mvidType));
             return _mvidField;
         }
 
+        internal Cci.IFieldReference GetOrAddInstrumentationPayloadRoot(int analysisKind, Cci.ITypeReference payloadRootType)
+        {
+            InstrumentationPayloadRootField payloadRootField;
+            if (!_instrumentationPayloadRootFields.TryGetValue(analysisKind, out payloadRootField))
+            {
+                Debug.Assert(!IsFrozen);
+                payloadRootField = _instrumentationPayloadRootFields.GetOrAdd(analysisKind, (kind) => new InstrumentationPayloadRootField(this, kind, payloadRootType));
```

Comment:
I am not sure if it is worth creating a lambda here. Since we already checked that key is not in the dictionary, simply passing a newly created object should be fine. There is a GetOrAdd overload like this.